scripts/src/java/org/oppia/android/scripts/proto/coverage.proto

```diff
@@ -0,0 +1,63 @@
+syntax = "proto3";
+
+package proto;
+
+option java_package = "org.oppia.android.scripts.proto";
+option java_multiple_files = true;
+
+message CoverageReport {
+  string bazel_test_target = 1;
+  repeated CoveredFile covered_file = 2;
```

Comment:
I'm curious: how valuable is it to store all of the covered files if we (generally) only care about a single source file per test target?

Separately, I wonder if we should model this report in the reverse order: have a single report for a source file and a list of test targets that were run for that _combined_ report. That better aligns with how we'll be using the protos, after all.

Reply:
Hey @BenHenning, I'm not entirely sure if I understand your question correctly. Are you asking why I'm storing all the covered files? 

Let's consider the scenario in **util/math** where we have several files like **RealExtensions.kt**, **FloatExtensions.kt**, **ExpressionToPolynomialConverter.kt**, **ExpressionToComparableOperationConverter.kt**, **ComparableOperationExtensions.kt**, and **others**. 

Typically, (_Bazel way of getting coverage_) when we perform coverage analysis, bazel generates a combined coverage report that includes data for every file in the package—that's the default behavior. 

However, in our case, we specifically want to gather coverage data for just a single file. 
**For example**, if we're interested in the coverage of **RealExtensions.kt**, the report should **only** **contain** RealExtensions.kt's coverage data, **disregarding others in the package**. 

This required approach was kept in mind and was implemented in PR 1.4 [[implementation](https://github.com/oppia/oppia-android/pull/5439/files#diff-e1015bdd1aafa12e3e95de3b408525733c702992881ece1d2399b57ded6a44fdR81-R89)], where it checks for the filename and parses the coverage data for the given filename while skipping the rest. This aligns with our original design intent.

Can I know if there are any conflicts with this approach or any concerns or if I've missed anything that needs to be addressed? Also, I'm not quite clear on what you mean by asking how valuable it is to store all covered files. 

In the later part where you mentioned about 'modelling this report in the reverse order' Could you please provide more details? Regarding your suggestion for a combined report, do you think it's necessary to have a combined report for all files in the packages? Should we proceed with that approach? Just wanted to know why so that could be helpful while generating custom reports!

Reply:
Let me try explaining my thoughts a different way, and it may be a bit repetitive to what you already said (apologies if that's the case).

The current modality of coverage in Bazel is to run coverage on one or more test targets. For parity with how we're using the script, let's assume one target. As you mentioned, Bazel will instrument all of the files under the instrumentation filter and should generate a single coverage.dat file that covers all of those files. To that end, it would make sense to design the proto to contain the tested target and all of the covered source files as part of the instrumentation filter.

However, how we want our coverage tool to work is to provide coverage analysis for only a single source file: the one that corresponds to the exact test target being checked. This begs two questions per the current design in this PR:
- There's really only one covered file we care about for each test target, so why have a list here?
- There's actually a many:1 mapping from target to source file (since a single source file may have a regular test, a local test, and a shared test). How do we consolidate these into a proper, single report for the original source file?

Separately, we also need to be able to collate the coverage results for each target/source file pair into a high-level report for eventual HTML/Markdown generation (since that step should be summarizing the code coverage for _all_ targets, not just a single one). How are we representing this part of the plan in the data structures being introduced here?

Does this help clarify?

Reply:
I think with this comment https://github.com/oppia/oppia-android/pull/5443#issuecomment-2189239096 and discussions things are clear, and we will need to possibly have the ability to run multiple targets/paths at once to replicate same wasy as to be done with CI for M2.